src/tactic/sine_filter.cpp

```diff
@@ -0,0 +1,252 @@
+    params_ref    m_params;
+    datatype_util m_dt;
+    ref<extension_model_converter> m_ext;
+    ref<filter_model_converter>    m_filter;
+    unsigned                       m_num_transformed;
+
+public:
+
+    sine_tactic(ast_manager& m, params_ref const& p): 
+        m(m), m_params(p), m_dt(m) {}
+    
+    virtual tactic * translate(ast_manager & m) {
+        return alloc(sine_tactic, m, m_params);
+    }
+
+    virtual void updt_params(params_ref const & p) {
+    }
+
+    virtual void collect_param_descrs(param_descrs & r) {
+    }
+
+    virtual void operator()(goal_ref const & g,
+                            goal_ref_buffer & result,
+                            model_converter_ref & mc,
+                            proof_converter_ref & pc,
+                            expr_dependency_ref & core) {
+        mc = 0; pc = 0; core = 0;
+
+        TRACE("sine", g->display(tout););
+        TRACE("sine", tout << g->size(););
+        ptr_vector<expr> new_forms;
+        filter_expressions(g, new_forms);
+        TRACE("sine", tout << new_forms.size(););
+        g->reset();
+        for (unsigned i = 0; i < new_forms.size(); i++) {
+          g->assert_expr(new_forms.get(i), 0, 0);
+        }
+        g->inc_depth();
+        g->updt_prec(goal::OVER);
+        result.push_back(g.get());
+        TRACE("sine", result[0]->display(tout););
+        SASSERT(g->is_well_sorted());
+        filter_model_converter * fmc = alloc(filter_model_converter, m);
+        mc = fmc;
+    }
+    
+    virtual void cleanup() {
+    }
+
+private:
+
+  // is this a user-defined symbol name?
+  bool is_name(func_decl * f) {
+    return f->get_family_id() < 0;
+  }
+
+  typedef std::pair<expr*,expr*> t_work_item;
+
+  t_work_item work_item(expr *e, expr *root) {
+    return std::pair<expr*, expr*>(e, root);
+  }
+
+  void find_constants(expr *e, obj_hashtable<func_decl> &consts) {
+    ptr_vector<expr> stack;
+    stack.push_back(e);
+    expr *curr;
+    while (!stack.empty()) {
+      curr = stack.back();
+      stack.pop_back();
+      if (is_app(curr)) {
+        app *a = to_app(curr);
+        func_decl *f = a->get_decl();
+        if (is_name(f) && !consts.contains(f)) {
+          consts.insert(f);
+        }
+      }
+    }
+  }
+
+  bool quantifier_matches(quantifier *q,
+                          obj_hashtable<func_decl> const & consts,
+                          ptr_vector<func_decl> & next_consts) {
+    TRACE("sine", tout << "size of consts is "; tout << consts.size(); tout << "\n";);
+    for (obj_hashtable<func_decl>::iterator constit = consts.begin(), constend = consts.end(); constit != constend; constit++) {
+      TRACE("sine", tout << *constit; tout << "\n";);
+    }
+    bool matched = false;
+    for (int i = 0; i < q->get_num_patterns(); i++) {
+      bool p_matched = true;
+      vector<expr *> stack;
+      expr *curr;
+      // patterns are wrapped with "pattern"
+      stack.push_back(to_app(q->get_pattern(i))->get_arg(0));
```

Comment:
we should add a function that wraps this to_app(q->get_pattern(i))->get_arg(0) and checks that indeed there is exactly one argument and that the function is "pattern". Since patterns are in the ast.h, the function would be a matcher on patterns. ast_manager has a method called "bool is_pattern(expr* e)". Patterns can be multi-patterns, so take more than 1 argument. The relevant matcher could then be bool is_pattern(expr* e, ptr_vector<expr>& args);.
You currently assume single patterns. Multi-patterns require matches on all argument terms (the set of patterns associated with a quantifier, on the other hand, are disjunctive: a match succeeds if one of the alternative patterns is matched).